CODING-CONVENTIONS.md

```diff
@@ -0,0 +1,226 @@
+  - Keep methods and classes short and focused on a single responsibility. Preferable maximum lengths:
+    * Lambdas: single line
+    * Methods: fit on a screen
+    * Anonymous classes: less than a screen
+    * Inner classes: not much more than a screen
+    * Classes: a few hundred lines
+  - Be thoughtfully organised in terms of method order, package structure, and module usage.
+  - Follow well-established patterns and conventions.
+  - Be consistent.
+  - Make it easy to follow the control flow by _clicking through_ in an IDE.
+  - Make it easier to use the right way than the wrong way. 
+  - Avoid abbreviations. We are a global team and when English is a second language abbreviations reduce readability. The following abbreviations are used: 
+    * tx -> Transaction
+    * IBFT -> Istanbul Byzantine Fault Tolerant (a consensus protocol)
+    * EVS -> Ethereum Virtual Machine 
+    
+# 3.2 Creating Code for Constant Refactoring and Evolving Design
+
+To be able to cope with constant refactoring and evolving design, write code that:
+* Is well tested. 
+
+  Avoid test cases requiring detailed interactions with mocks because these are often brittle.
+
+* Avoids duplication.
+
+* Follows the Single Responsibility Principle where each class is responsible for one thing.
+
+  >**Note** It is important to scope the responsibility wisely. Responsibilities that are:
+  > *  Too small lead to an explosion of classes making things hard to follow
+  > *  Too large lead to the class becoming big and unwieldy.
+
+* Favors composition over inheritance. You can use inheritance, but prefer composition.
+
+* Uses dependency injection
+  - Constructors should be simple, with dependencies passed in rather than built in the constructor.
+  - Pantheon does not use a dependency injection framework.
+
+* Validates method parameters for public methods using the Guava `Preconditions` class. Avoid validating parameters in private methods.
+
+* Generally avoids interfaces with a single implementation unless they are explicitly being used to narrow the exposed API.
+
+* Uses the most general applicable type. For example,  `List` or `Collection` instead of `ArrayList`.
+
+## 3.3 Additional Design Elements
+
+Additional design elements include: 
+* Using Optional rather than returning null when not having a value is a normal case.
+* Considering exception and error handling as part of the overall design.   Pantheon avoids checked exceptions.
+* Giving threads are giving meaningful names. For example:
+ `Executors.newFixedThreadPool(1, new ThreadFactoryBuilder().setNameFormat(“Ibft”).build())`
+
+  
+# 4 Java 
+
+## 4.1 Style Guide 
+
+Pantheon follows the [Google code style](https://google.github.io/styleguide/javaguide.html) and uses spotless to ensure consistency of formatting. 
+
+To automatically reformat the code before creating a pull request, run: 
+
+```json
+./gradlew spotlessApply
+```
+
+### 4.1.1 Install Google Style Settings 
```

Comment:
This is out of date.  We use google-java-format in spotless so I think we should recommend it for the style guide.  IntelliJ instructions are here - https://github.com/google/google-java-format#intellij and eclipse here - https://github.com/google/google-java-format#eclipse.  NetBeans and visual studio code is out of luck

Reply:
Fixed.